File: src/psm-core/src/indexables.ts
import { tokenize } from "./ranking.js";
import type { IndexableKind, IndexablePayload, IndexableRecord, MemoryFactPayload } from "./types.js";

const WORKFLOW_HEADER_PATTERNS: Array<{ pattern: RegExp; key: string }> = [
  { pattern: /review.*pull request|pull request review/i, key: "review-pr" },
  { pattern: /runpod.*train|gpu train|train.*runpod/i, key: "runpod-gpu-train" },
  { pattern: /grounding bar|promotion bar/i, key: "grounding-bar" }
];

export interface BuildIndexablesInput {
  llmResponse: string;
  memoryContent: string;
  tags?: string[];
  memoryTable?: string;
  memoryId?: string;
  facts?: MemoryFactPayload[];
  explicitIndexables?: IndexablePayload[];
}

export function buildIndexablesForRemember(input: BuildIndexablesInput): IndexablePayload[] {
  if (input.explicitIndexables?.length) {
    return input.explicitIndexables
      .map((row) => normalizeIndexable(row, input))
      .filter((row): row is IndexablePayload => row !== null);
  }

  const sourceText = input.llmResponse.trim() || input.memoryContent.trim();
  if (!sourceText) return [];

  const workflowKey = inferWorkflowKey(sourceText, input.tags ?? []);
  const steps = extractWorkflowSteps(sourceText);
  if (workflowKey && steps.length >= 2) {
    return [{
      kind: "workflow",
      key: workflowKey,
      target_memory_table: input.memoryTable,
      target_memory_id: input.memoryId,
      steps,
      salience: 0.95,
      reconstructive_hint: reconstructiveHint(sourceText),
      evidence_text: sourceText.slice(0, 500),
      tags: uniqueTags([`workflow:${workflowKey}`, "workflow", ...(input.tags ?? [])])
    }];
  }

  const content = cleanText(input.memoryContent || sourceText);
  const mnemonicKey = buildMnemonicKey(content, input.tags ?? []);
  const rows: IndexablePayload[] = [{
    kind: "mnemonic",
    key: mnemonicKey,
    target_memory_table: input.memoryTable,
    target_memory_id: input.memoryId,
    salience: salienceFor(content, input.tags ?? []),
    reconstructive_hint: reconstructiveHint(content),
    evidence_text: content,
    tags: uniqueTags(input.tags ?? []).slice(0, 6)
  }];

  const factKey = buildFactAnchorKey(input.facts ?? []);
  if (factKey && factKey !== mnemonicKey) {
    rows.push({
      kind: "fact_anchor",
      key: factKey,
      target_memory_table: input.memoryTable,
      target_memory_id: input.memoryId,
      salience: Math.max(rows[0].salience ?? 0.8, 0.82),
      reconstructive_hint: reconstructiveHint(content),
      evidence_text: content,
      tags: uniqueTags(input.tags ?? []).slice(0, 6)
    });
  }
  return rows;
}

export function inferWorkflowKey(text: string, tags: string[] = []): string | null {
  for (const tag of tags) {
    const match = String(tag).match(/^workflow:([a-z0-9-]+)$/i);
    if (match) return match[1].toLowerCase();
  }
  const header = text.match(/^#\s+(.+)$/m)?.[1] ?? "";
  const haystack = `${header}\n${text.slice(0, 240)}`;
  for (const entry of WORKFLOW_HEADER_PATTERNS) {
    if (entry.pattern.test(haystack)) return entry.key;
  }
  return null;
}

export function extractWorkflowSteps(text: string): string[] {
  const steps: string[] = [];
  for (const line of text.split("\n")) {
    const match = line.match(/^\s*\d+\.\s+(.+?)\s*$/);
    if (!match?.[1]) continue;
    steps.push(stepToId(match[1]));
  }
  return unique(steps);
}

export function normalizeRecallKey(query: string): string {
  return query
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

export interface ScoredIndexable extends IndexableRecord {
  score: number;
}

export function rankIndexables(query: string, indexables: IndexableRecord[], topK = 5): ScoredIndexable[] {
  const normalized = normalizeRecallKey(query);
  const tokens = tokenize(query);
  return indexables
    .map((row) => ({ ...row, score: indexableScore(normalized, tokens, row) }))
    .filter((row) => row.score >= 0.35)
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);
}

function indexableScore(normalizedQuery: string, queryTokens: string[], row: IndexableRecord): number {
  let score = row.salience ?? 0.5;
  const key = row.key.toLowerCase();
  if (key && (normalizedQuery === key || normalizedQuery.includes(key) || key.includes(normalizedQuery))) {
    score += 0.55;
  }
  if (row.kind === "workflow" && /review|workflow|procedure|how do i/i.test(normalizedQuery.replace(/-/g, " "))) {
    score += 0.1;
  }
  const searchable = [row.key, row.reconstructive_hint ?? "", ...(row.tags ?? []), ...(row.steps ?? [])].join(" ");
  const haystack = new Set(tokenize(searchable));
  const overlap = queryTokens.filter((token) => haystack.has(token)).length;
  if (queryTokens.length > 0) {
    score += overlap / queryTokens.length * 0.35;
  }
  return Number(score.toFixed(6));
}

function normalizeIndexable(row: IndexablePayload, input: BuildIndexablesInput): IndexablePayload | null {
  const key = cleanKey(row.key);
  if (!key) return null;
  return {
    kind: row.kind ?? "mnemonic",
    key,
    target_memory_table: row.target_memory_table ?? input.memoryTable,
    target_memory_id: row.target_memory_id ?? input.memoryId,
    steps: Array.isArray(row.steps) ? row.steps.map((step) => stepToId(String(step))) : undefined,
    salience: clamp01(row.salience ?? 0.85),
    reconstructive_hint: row.reconstructive_hint ?? reconstructiveHint(input.memoryContent),
    evidence_text: row.evidence_text ?? input.memoryContent,
    tags: uniqueTags(row.tags ?? input.tags ?? [])
  };
}

function buildMnemonicKey(content: string, tags: string[]): string {
  const tagTokens = tags.flatMap((tag) => meaningfulTokens(String(tag).replace(/_/g, " ")));
  const contentTokens = meaningfulTokens(content);
  const tokens = unique([...tagTokens, ...contentTokens]).slice(0, 4);
  return tokens.length > 0 ? tokens.join("-") : "memory-anchor";
}

function buildFactAnchorKey(facts: MemoryFactPayload[]): string {
  const fact = facts.find((item) => item.subject && item.predicate && item.value_text);
  if (!fact) return "";
  return unique([
    ...meaningfulTokens(fact.subject ?? ""),
    ...meaningfulTokens(fact.predicate ?? ""),
    ...meaningfulTokens(fact.value_text ?? "")
  ]).slice(0, 4).join("-");
}

function stepToId(step: string): string {
  return step
    .toLowerCase()
    .replace(/`([^`]+)`/g, "$1")
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "")
    .slice(0, 48) || "step";
}

function salienceFor(content: string, tags: string[]): number {
  const lower = content.toLowerCase();
  let score = 0.68;
  if (/\b\d{4}\b|yesterday|last week|workflow|review|procedure/.test(lower)) score += 0.08;
  if (/decision|prefer|constraint|indexable|mnemonic|recall/.test(lower)) score += 0.12;
  if (tags.length > 0) score += 0.04;
  return Number(Math.min(score, 0.98).toFixed(2));
}

function reconstructiveHint(content: string): string {
  const sentence = content.match(/^(.+?[.!?])(?:\s|$)/)?.[1] ?? content;
  return sentence.length <= 160 ? sentence : `${sentence.slice(0, 157).trim()}...`;
}

function meaningfulTokens(text: string): string[] {
  const stop = new Set(["the", "and", "for", "that", "this", "with", "from", "into", "said", "user", "memory"]);
  return cleanText(text)
    .toLowerCase()
    .match(/[a-z0-9]+/g)
    ?.filter((token) => token.length > 2 && !stop.has(token)) ?? [];
}

function cleanText(value: string): string {
  return value.trim().replace(/\s+/g, " ");
}

function cleanKey(value: string): string {
  return value.trim().toLowerCase().replace(/[^a-z0-9-]+/g, "-").replace(/^-+|-+$/g, "");
}

function uniqueTags(tags: string[]): string[] {
  return [...new Set(tags.map((tag) => String(tag).trim()).filter(Boolean).map((tag) => tag.replace(/\s+/g, "_")))];
}

function unique<T>(values: T[]): T[] {
  return [...new Set(values)];
}

function clamp01(value: number): number {
  return Math.max(0, Math.min(1, value));
}

File: dotnet/src/PsmMemory.Core/Indexables.cs
using System.Text.RegularExpressions;
using PsmMemory.Core.Models;

namespace PsmMemory.Core;

/// <summary>
/// Ported directly from psm-core/src/indexables.ts: buildIndexablesForRemember() and its helpers,
/// plus rankIndexables()/normalizeRecallKey(). Pure deterministic synthesis -- no LLM call involved
/// (see PsmService.RememberAsync remarks for the call site: this only ever runs when a storage
/// decision would actually write a memory and didn't already carry explicit indexables). Style
/// mirrors <see cref="TextSegmenter"/>: a static partial class of source-generated regexes plus
/// small deterministic string-transform helpers.
/// </summary>
public static partial class Indexables
{
    // Exact hardcoded table from indexables.ts's WORKFLOW_HEADER_PATTERNS -- do not add entries
    // without a corresponding change in the original TS source; this port must stay 1:1 with it.
    [GeneratedRegex(@"review.*pull request|pull request review", RegexOptions.IgnoreCase)]
    private static partial Regex ReviewPrRegex();

    [GeneratedRegex(@"runpod.*train|gpu train|train.*runpod", RegexOptions.IgnoreCase)]
    private static partial Regex RunpodTrainRegex();

    [GeneratedRegex(@"grounding bar|promotion bar", RegexOptions.IgnoreCase)]
    private static partial Regex GroundingBarRegex();

    [GeneratedRegex(@"^workflow:([a-z0-9-]+)$", RegexOptions.IgnoreCase)]
    private static partial Regex WorkflowTagRegex();

    // Deliberately a single '#' (not '#{1,3}' like TextSegmenter's header splitter) -- indexables.ts
    // only treats a top-level H1 as a workflow title.
    [GeneratedRegex(@"^#\s+(.+)$", RegexOptions.Multiline)]
    private static partial Regex H1HeaderRegex();

    [GeneratedRegex(@"^\s*\d+\.\s+(.+?)\s*$")]
    private static partial Regex NumberedStepRegex();

    [GeneratedRegex("`([^`]+)`")]
    private static partial Regex BacktickRegex();

    [GeneratedRegex("[^a-z0-9]+")]
    private static partial Regex NonAlnumRunRegex();

    [GeneratedRegex("[^a-z0-9-]+")]
    private static partial Regex NonAlnumOrHyphenRunRegex();

    [GeneratedRegex("[a-z0-9]+")]
    private static partial Regex AlnumTokenRegex();

    [GeneratedRegex(@"\s+")]
    private static partial Regex WhitespaceRunRegex();

    [GeneratedRegex(@"^(.+?[.!?])(?:\s|$)")]
    private static partial Regex FirstSentenceRegex();

    [GeneratedRegex(@"\b\d{4}\b|yesterday|last week|workflow|review|procedure")]
    private static partial Regex SalienceSituationalRegex();

    [GeneratedRegex(@"decision|prefer|constraint|indexable|mnemonic|recall")]
    private static partial Regex SalienceImportanceRegex();

    [GeneratedRegex(@"review|workflow|procedure|how do i", RegexOptions.IgnoreCase)]
    private static partial Regex WorkflowQueryRegex();

    // A separate stopword list from Ranking.Stopwords by design (see task brief) -- meaningfulTokens
    // in the original TS never stems tokens the way ranking.ts's tokenize() does, and the two lists
    // were tuned independently for different purposes (key-building vs. relevance ranking).
    private static readonly HashSet<string> MeaningfulTokenStopwords = new()
    {
        "the", "and", "for", "that", "this", "with", "from", "into", "said", "user", "memory"
    };

    /// <summary>Ported from indexables.ts's BuildIndexablesInput.</summary>
    public sealed class BuildIndexablesInput
    {
        public required string LlmResponse { get; init; }
        public required string MemoryContent { get; init; }
        public List<string>? Tags { get; init; }
        public string? MemoryTable { get; init; }
        public string? MemoryId { get; init; }
        public List<MemoryFactPayload>? Facts { get; init; }
        public List<IndexablePayload>? ExplicitIndexables { get; init; }
    }

    /// <summary>An <see cref="IndexableRecord"/> scored against a query by <see cref="RankIndexables"/>.</summary>
    public sealed class ScoredIndexable
    {
        public required IndexableRecord Record { get; init; }
        public required double Score { get; init; }

        public string Id => Record.Id;
        public string Kind => Record.Kind;
        public string Key => Record.Key;
        public string? TargetMemoryTable => Record.TargetMemoryTable;
        public string? TargetMemoryId => Record.TargetMemoryId;
        public List<string> Steps => Record.Steps;
        public double Salience => Record.Salience;
        public string? ReconstructiveHint => Record.ReconstructiveHint;
        public string? EvidenceText => Record.EvidenceText;
        public List<string> Tags => Record.Tags;
    }

    /// <summary>
    /// Ported directly from indexables.ts's buildIndexablesForRemember(). Deterministic, no LLM
    /// call: (1) explicit indexables the model already emitted just get normalized/passed through;
    /// (2) otherwise, a workflow row is synthesized when the source text has BOTH a recognized
    /// workflow header/tag AND &gt;=2 numbered steps; (3) otherwise a mnemonic row is always produced,
    /// plus an additional fact_anchor row when the supplied facts key off to something distinct from
    /// the mnemonic key.
    /// </summary>
    public static List<IndexablePayload> BuildIndexablesForRemember(BuildIndexablesInput input)
    {
        if (input.ExplicitIndexables is { Count: > 0 })
        {
            var normalized = new List<IndexablePayload>();
            foreach (var row in input.ExplicitIndexables)
            {
                var result = NormalizeIndexable(row, input);
                if (result is not null) normalized.Add(result);
            }
            return normalized;
        }

        var sourceText = input.LlmResponse.Trim();
        if (sourceText.Length == 0) sourceText = input.MemoryContent.Trim();
        if (sourceText.Length == 0) return new List<IndexablePayload>();

        var tags = input.Tags ?? new List<string>();

        var workflowKey = InferWorkflowKey(sourceText, tags);
        var steps = ExtractWorkflowSteps(sourceText);
        if (workflowKey is not null && steps.Count >= 2)
        {
            return new List<IndexablePayload>
            {
                new()
                {
                    Kind = IndexableKinds.Workflow,
                    Key = workflowKey,
                    TargetMemoryTable = input.MemoryTable,
                    TargetMemoryId = input.MemoryId,
                    Steps = steps,
                    Salience = 0.95,
                    ReconstructiveHint = ReconstructiveHint(sourceText),
                    EvidenceText = sourceText.Length > 500 ? sourceText[..500] : sourceText,
                    Tags = UniqueTags(new[] { $"workflow:{workflowKey}", "workflow" }.Concat(tags))
                }
            };
        }

        var content = CleanText(!string.IsNullOrEmpty(input.MemoryContent) ? input.MemoryContent : sourceText);
        var mnemonicKey = BuildMnemonicKey(content, tags);
        var rows = new List<IndexablePayload>
        {
            new()
            {
                Kind = IndexableKinds.Mnemonic,
                Key = mnemonicKey,
                TargetMemoryTable = input.MemoryTable,
                TargetMemoryId = input.MemoryId,
                Salience = SalienceFor(content, tags),
                ReconstructiveHint = ReconstructiveHint(content),
                EvidenceText = content,
                Tags = UniqueTags(tags).Take(6).ToList()
            }
        };

        var factKey = BuildFactAnchorKey(input.Facts);
        if (factKey.Length > 0 && factKey != mnemonicKey)
        {
            rows.Add(new IndexablePayload
            {
                Kind = IndexableKinds.FactAnchor,
                Key = factKey,
                TargetMemoryTable = input.MemoryTable,
                TargetMemoryId = input.MemoryId,
                Salience = Math.Max(rows[0].Salience ?? 0.8, 0.82),
                ReconstructiveHint = ReconstructiveHint(content),
                EvidenceText = content,
                Tags = UniqueTags(tags).Take(6).ToList()
            });
        }

        return rows;
    }

    /// <summary>Ported from indexables.ts's inferWorkflowKey().</summary>
    public static string? InferWorkflowKey(string text, IReadOnlyList<string>? tags)
    {
        foreach (var tag in tags ?? Array.Empty<string>())
        {
            var match = WorkflowTagRegex().Match(tag);
            if (match.Success) return match.Groups[1].Value.ToLowerInvariant();
        }

        var headerMatch = H1HeaderRegex().Match(text);
        var header = headerMatch.Success ? headerMatch.Groups[1].Value : "";
        var haystack = header + "\n" + (text.Length > 240 ? text[..240] : text);

        if (ReviewPrRegex().IsMatch(haystack)) return "review-pr";
        if (RunpodTrainRegex().IsMatch(haystack)) return "runpod-gpu-train";
        if (GroundingBarRegex().IsMatch(haystack)) return "grounding-bar";
        return null;
    }

    /// <summary>Ported from indexables.ts's extractWorkflowSteps().</summary>
    public static List<string> ExtractWorkflowSteps(string text)
    {
        var steps = new List<string>();
        foreach (var line in text.Split('\n'))
        {
            var match = NumberedStepRegex().Match(line);
            if (!match.Success || match.Groups[1].Value.Length == 0) continue;
            steps.Add(StepToId(match.Groups[1].Value));
        }
        return Unique(steps);
    }

    /// <summary>Ported from indexables.ts's stepToId().</summary>
    public static string StepToId(string step)
    {
        var lowered = step.ToLowerInvariant();
        var noBackticks = BacktickRegex().Replace(lowered, "$1");
        var cleaned = NonAlnumRunRegex().Replace(noBackticks, "_").Trim('_');
        var truncated = cleaned.Length > 48 ? cleaned[..48] : cleaned;
        return truncated.Length > 0 ? truncated : "step";
    }

    /// <summary>Ported from indexables.ts's normalizeRecallKey().</summary>
    public static string NormalizeRecallKey(string query)
    {
        var lowered = query.ToLowerInvariant();
        var replaced = NonAlnumRunRegex().Replace(lowered, "-");
        return replaced.Trim('-');
    }

    /// <summary>
    /// Ported directly from indexables.ts's rankIndexables(). Called only from the recall() call
    /// path -- context() never ranks/surfaces indexables in the original TS source (confirmed via
    /// grep: no `rankIndexables` call anywhere in context()).
    /// </summary>
    public static List<ScoredIndexable> RankIndexables(string query, IEnumerable<IndexableRecord> indexables, int topK = 5)
    {
        var normalized = NormalizeRecallKey(query);
        var tokens = Ranking.Tokenize(query);
        return indexables
            .Select(row => new ScoredIndexable { Record = row, Score = IndexableScore(normalized, tokens, row) })
            .Where(scored => scored.Score >= 0.35)
            .OrderByDescending(scored => scored.Score)
            .Take(topK)
            .ToList();
    }

    /// <summary>Ported from indexables.ts's indexableScore().</summary>
    private static double IndexableScore(string normalizedQuery, List<string> queryTokens, IndexableRecord row)
    {
        var score = row.Salience;
        var key = row.Key.ToLowerInvariant();
        if (key.Length > 0 && (normalizedQuery == key || normalizedQuery.Contains(key) || key.Contains(normalizedQuery)))
        {
            score += 0.55;
        }
        if (row.Kind == IndexableKinds.Workflow && WorkflowQueryRegex().IsMatch(normalizedQuery.Replace("-", " ")))
        {
            score += 0.1;
        }
        var searchable = string.Join(" ", new[] { row.Key, row.ReconstructiveHint ?? "" }.Concat(row.Tags).Concat(row.Steps));
        var haystack = new HashSet<string>(Ranking.Tokenize(searchable));
        var overlap = queryTokens.Count(t => haystack.Contains(t));
        if (queryTokens.Count > 0) score += (double)overlap / queryTokens.Count * 0.35;
        return Math.Round(score, 6);
    }

    /// <summary>Ported from indexables.ts's normalizeIndexable() (the explicitIndexables branch).</summary>
    private static IndexablePayload? NormalizeIndexable(IndexablePayload row, BuildIndexablesInput input)
    {
        var key = CleanKey(row.Key);
        if (key.Length == 0) return null;
        return new IndexablePayload
        {
            Kind = string.IsNullOrEmpty(row.Kind) ? IndexableKinds.Mnemonic : row.Kind,
            Key = key,
            TargetMemoryTable = row.TargetMemoryTable ?? input.MemoryTable,
            TargetMemoryId = row.TargetMemoryId ?? input.MemoryId,
            Steps = row.Steps?.Select(StepToId).ToList(),
            Salience = Clamp01(row.Salience ?? 0.85),
            ReconstructiveHint = row.ReconstructiveHint ?? ReconstructiveHint(input.MemoryContent),
            EvidenceText = row.EvidenceText ?? input.MemoryContent,
            Tags = UniqueTags(row.Tags ?? input.Tags ?? new List<string>())
        };
    }

    /// <summary>Ported from indexables.ts's buildMnemonicKey().</summary>
    private static string BuildMnemonicKey(string content, IReadOnlyList<string> tags)
    {
        var tagTokens = tags.SelectMany(tag => MeaningfulTokens(tag.Replace("_", " ")));
        var contentTokens = MeaningfulTokens(content);
        var tokens = Unique(tagTokens.Concat(contentTokens)).Take(4).ToList();
        return tokens.Count > 0 ? string.Join("-", tokens) : "memory-anchor";
    }

    /// <summary>Ported from indexables.ts's buildFactAnchorKey().</summary>
    private static string BuildFactAnchorKey(IReadOnlyList<MemoryFactPayload>? facts)
    {
        var fact = (facts ?? Array.Empty<MemoryFactPayload>()).FirstOrDefault(f =>
            !string.IsNullOrEmpty(f.Subject) && !string.IsNullOrEmpty(f.Predicate) && !string.IsNullOrEmpty(f.ValueText));
        if (fact is null) return "";

        var tokens = Unique(
                MeaningfulTokens(fact.Subject ?? "")
                    .Concat(MeaningfulTokens(fact.Predicate ?? ""))
                    .Concat(MeaningfulTokens(fact.ValueText ?? "")))
            .Take(4)
            .ToList();
        return string.Join("-", tokens);
    }

    /// <summary>Ported from indexables.ts's salienceFor().</summary>
    private static double SalienceFor(string content, IReadOnlyList<string> tags)
    {
        var lower = content.ToLowerInvariant();
        var score = 0.68;
        if (SalienceSituationalRegex().IsMatch(lower)) score += 0.08;
        if (SalienceImportanceRegex().IsMatch(lower)) score += 0.12;
        if (tags.Count > 0) score += 0.04;
        return Math.Round(Math.Min(score, 0.98), 2);
    }

    /// <summary>Ported from indexables.ts's reconstructiveHint().</summary>
    private static string ReconstructiveHint(string content)
    {
        var match = FirstSentenceRegex().Match(content);
        var sentence = match.Success ? match.Groups[1].Value : content;
        return sentence.Length <= 160 ? sentence : sentence[..157].Trim() + "...";
    }

    /// <summary>Ported from indexables.ts's meaningfulTokens(). Deliberately does NOT stem tokens
    /// the way Ranking.Tokenize does -- the original TS never shared that logic between the two.</summary>
    private static List<string> MeaningfulTokens(string text)
    {
        var cleaned = CleanText(text).ToLowerInvariant();
        var result = new List<string>();
        foreach (Match m in AlnumTokenRegex().Matches(cleaned))
        {
            if (m.Value.Length > 2 && !MeaningfulTokenStopwords.Contains(m.Value)) result.Add(m.Value);
        }
        return result;
    }

    /// <summary>Ported from indexables.ts's cleanText().</summary>
    private static string CleanText(string value) => WhitespaceRunRegex().Replace(value.Trim(), " ");

    /// <summary>Ported from indexables.ts's cleanKey().</summary>
    private static string CleanKey(string value) =>
        NonAlnumOrHyphenRunRegex().Replace(value.Trim().ToLowerInvariant(), "-").Trim('-');

    /// <summary>Ported from indexables.ts's uniqueTags().</summary>
    private static List<string> UniqueTags(IEnumerable<string> tags)
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var raw in tags)
        {
            var trimmed = raw?.Trim();
            if (string.IsNullOrEmpty(trimmed)) continue;
            var normalized = WhitespaceRunRegex().Replace(trimmed, "_");
            if (seen.Add(normalized)) result.Add(normalized);
        }
        return result;
    }

    private static List<string> Unique(IEnumerable<string> values)
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var v in values)
        {
            if (seen.Add(v)) result.Add(v);
        }
        return result;
    }

    private static double Clamp01(double value) => Math.Max(0, Math.Min(1, value));
}
